### PR TITLE
#402 Remove intermediary switchOff variable in Network CPP bus model …

### PR DESCRIPTION
**Pull request by rosiereflo**
…and add a routine during the model initialization to propagate the potential modifications of discrete variables during the initialization (i.e. opening buses if several subnetworks exist)

### CODE REVIEW

File: dynawo/sources/Modeler/Common/DYNModelMulti.cpp [hunk 1]
```diff
@@ -248,24 +248,66 @@ ModelMulti::initBuffers() {
...
 
+  // Detect if some discrete variable was modified during the initialization (e.g. subnetwork detection)
+  std::vector<double> z(sizeZ());
+  z.assign(zLocal_, zLocal_ + sizeZ());
+  vector<int> indicesDiff;
+  for (unsigned int i = 0; i < z.size(); ++i) {
+    if (doubleNotEquals(z[i], zSave_[i])) {
+      indicesDiff.push_back(i);
+    }
+  }
```
Comment (adriguir) [inline dynawo/sources/Modeler/Common/DYNModelMulti.cpp]: At first glance, it is not obvious to understand why you need to introduce this additional vector for z - I finally see that you use it below as a kind of save vector.

Shouldn't it be possible to only have zSave_ and zLocal_?

It leads to something like:
...
vector<int> indicesDiff;
for (unsigned int i = 0; i < sizeZ(); ++i) {
   if (doubleNotEquals(zLocal_[i], zSave_[i])) {
        indicesDiff...
}
}
zSave_.assign(zLocal_, zLocal_ + sizeZ());

File: dynawo/sources/Modeler/Common/DYNModelMulti.cpp [hunk 2]
```diff
@@ -248,24 +248,66 @@ ModelMulti::initBuffers() {
...
+    for (unsigned i = 0; i < 10 && !vectorAreEquals(zSave_, z); ++i) {
+      copieResultZ(z);
+      std::copy(z.begin(), z.end(), zLocal_);
+      for (unsigned int i = 0; i < subModels_.size(); ++i)
+        subModels_[i]->evalGSub(t0);
+      for (unsigned int i = 0; i < subModels_.size(); ++i)
+        subModels_[i]->evalZSub(t0);
+      z.assign(zLocal_, zLocal_ + sizeZ());
+    }
+  }
```
Comment (adriguir) [inline dynawo/sources/Modeler/Common/DYNModelMulti.cpp]: Same comment here: possible only with zSave_ and zLocal_? Avoiding some copy?

File: dynawo/sources/Modeler/Common/DYNModelMulti.cpp (hunk 1 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNModelMulti.cpp]: This intermediary vector was used to be able to use the copieResultZ method to be as close as possible of evalZ

File: dynawo/sources/Modeler/Common/DYNModelMulti.cpp [hunk 3]
```diff
@@ -248,24 +248,66 @@ ModelMulti::initBuffers() {
 void
 ModelMulti::init(const double& t0) {
   Timer timer1("ModelMulti::init");
+
+  zSave_.resize(sizeZ(), 0.);
+  zSave_.assign(zLocal_, zLocal_ + sizeZ());
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/Common/DYNModelMulti.cpp]: Solved

**Pull request merged**